packages/select/src/Select.tsx

```diff
@@ -0,0 +1,470 @@
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+import PropTypes from 'prop-types';
+import { cx, css } from '@leafygreen-ui/emotion';
+import { IdAllocator } from '@leafygreen-ui/lib';
+import { breakpoints } from '@leafygreen-ui/tokens';
+import { colorSets, mobileSizeSet, Mode, Size, sizeSets } from './styleSets';
+import ListMenu from './ListMenu';
+import MenuButton from './MenuButton';
+import SelectContext from './SelectContext';
+import { InternalOption, OptionElement } from './Option';
+import {
+  convertToInternalElements,
+  getOptionValue,
+  isOptionDisabled,
+  isOptionSelectable,
+  reconcileOption,
+  traverseSelectChildren,
+  useSmartRef,
+} from './utils';
+import { useViewportSize } from '@leafygreen-ui/hooks/src';
+
+const labelStyle = css`
+  display: inline-block;
+  margin-bottom: 2px;
+  font-weight: bold;
+`;
+
+export type Props = {
+  children: React.ReactNode;
+  className?: string;
+  id?: string;
+  darkMode?: boolean;
+  size?: Size;
+  disabled?: boolean;
+  label: string;
+  description?: string;
+  placeholder?: string;
+  name: string;
+} & (
+  | // Uncontrolled
+  ({
+      defaultValue?: string;
+      value?: undefined;
+    } & {
+      onChange?: (value: string) => void;
+      readOnly?: false;
+    })
+  // Controlled
+  | ({ value: string; defaultValue?: undefined } & (
+      | {
+          onChange: (value: string) => void;
+          readOnly?: false;
+        }
+      | { readOnly: true; onChange?: undefined }
+    ))
+);
+
+const idAllocator = IdAllocator.create('select');
+
+export default function Select({
+  children,
+  darkMode = false,
+  size = Size.Normal,
+  disabled = false,
+  className,
+  id: idProp,
+  label,
+  description,
+  placeholder = 'Select',
+  name,
+  defaultValue,
+  value,
+  onChange,
+  readOnly,
+}: Props) {
+  const id = useMemo(() => idProp ?? idAllocator.generate(), [idProp]);
+  const labelId = `${id}-label`;
+  const descriptionId = `${id}-description`;
+  const menuId = `${id}-menu`;
+
+  const [open, setOpen] = useState(false);
+
+  const menuButtonRef = useSmartRef<HTMLDivElement | null>(null);
```

Comment:
for my own understanding, what's the benefit of this over useElementNode?

Reply:
```tsx
function Parent(props) {
  const [ref, setRef] = useElementNode();

  return <Child ref={setRef} />;
}

function Child({ref: setRef}) {
  const ref = useRef();
  return <div ref={/* how to set both `setRef` and `ref`? */} />;
}
```

```tsx
function Parent(props) {
  const ref = useSmartRef();

  return <Child ref={ref} />;
}

function Child({ref}) {
  // can just access `ref.current`

  return <div ref={ref} />;
}
```

Reply:
Ah! Neat!